### PR TITLE
Revert "[bundle] Put the Gemfile on a diet and load gems on-demand"

### PR DESCRIPTION
**Pull request by yuki24**
As I described in the previous thread, this would break a lot of conventions Rails developers are familiar with. In my experience, changes that get a Rails app off of "Rails" would always bite us later on.

 * Changing how the app starts up would be surprising for other Rails developers
 * Having a different setup would make framework upgrades harder
 * Rails engines written based on Rails' conventions might not work due to the setup differences

I also wouldn't feel comfortable having to require dependencies explicitly on Rails apps.

If booting up time is a problem, I would suggest using [zeus](https://github.com/burke/zeus) or [bootsnap](https://github.com/Shopify/bootsnap).

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,42 +1,49 @@
...
-  gem 'puma'
-end
 
-group :app do
-  gem 'email_validator'
-  gem 'paper_trail'
-  gem 'paper_trail-globalid'
-end
+# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
+gem 'rails', '~> 5.1.1'
```
Comment (yuki24) [inline Gemfile]: This'd bring back `ActionCable` but we can remove that once we merge this PR.

File: Gemfile [hunk 2]
```diff
@@ -1,42 +1,49 @@
...
+
+gem 'paper_trail'
+gem 'paper_trail-globalid'
+
+gem 'email_validator', require: 'email_validator/strict'
+
+group :development, :test do
+  # Call 'byebug' anywhere in the code to stop execution and get a debugger console
+  gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
+  gem 'rspec-rails'
```
Comment (yuki24) [inline Gemfile]: Also, `rspec-rails` can go to just `:test`.

File: Gemfile [hunk 3]
```diff
@@ -1,42 +1,49 @@
...
 
-group :development, :test do
-  gem 'byebug'
-  gem 'rspec-rails'
+group :development do
+  gem 'annotate'
 end
+
+# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
+gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
```
Comment (yuki24) [inline Gemfile]: This can also be removed, but I'll send a separate PR.

**Comment by alloy** [comment]
I left a response on the original thread https://github.com/artsy/constellation/pull/11#issuecomment-302703734.

I’m against reverting this or adding more tooling at the stage.

**Comment by alloy** [comment]
I’m going to close this for now. I hear your concerns, but I think that one important outcome we both want is the same, namely being able to work on actual product features and not dealing with tech details anymore than necessary. Please do revisit this when you run into a case where you think it’s hindering 🙏

**Pull request closed without merging**